tools/shared/src/main/scala/org/scalajs/core/tools/javascript/ScalaJSClassEmitter.scala

```diff
@@ -264,8 +267,25 @@ final class ScalaJSClassEmitter private (
       val js.Function(params, body) = genConstructorFunForJSClass(tree)
       js.MethodDef(static = false, js.Ident("constructor"), params, body)
     } else {
+      val className = tree.name.name
       val fieldDefs = genFieldDefs(tree)
-      if (fieldDefs.isEmpty && outputMode == OutputMode.ECMAScript6) {
+
+      val (otherStatements, args) = if (usesJSConstructorOpt(className)) {
+        val (head :: _) =
+          tree.memberMethods.map(_.tree).
+          filter(x => isConstructorName(x.name.name))
+        val js.Function(args, bodyBlock) = desugarToConstructor(this,
+            tree.name.name, head.args, head.body, js.Skip())
+        val body = bodyBlock match {
+          case js.Block(bdy) => bdy
+          case _             => Nil
```

Comment:
With my comment below, you could just as always return `bodyBlock` from the then branch, and rename it as `ctorBody`.